Cocoa WorkQueue::dispatchWithQOS() and dispatch() work item lifetime semantics do not match

kkinnunen-apple opened the pull request:
#### 56fa22f9af0ff3e8bb475e9293bf8da34d88926b
<pre>
Cocoa WorkQueue::dispatchWithQOS() and dispatch() work item lifetime semantics do not match
<a href="https://bugs.webkit.org/show_bug.cgi?id=260852">https://bugs.webkit.org/show_bug.cgi?id=260852</a>
rdar://114618174

Reviewed by Jean-Yves Avenard and Chris Dumez.

dispatch() users expect the work function to be destructed in the
queue

Add the same semantic to dispatchWithQOS() to avoid confusion in the
code and to have uniform interface.

* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::dispatchWithQOS):
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/267776@main">https://commits.webkit.org/267776@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da4fa1140311b425af3c1637118ebbff72a7821e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17545 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18552 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15708 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16959 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18003 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14516 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19360 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15200 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21960 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14464 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15569 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19686 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15998 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15970 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13545 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18328 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15145 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4286 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4208 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19510 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19551 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15801 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4136 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
```diff
@@ -306,8 +306,78 @@ TEST(WTF_WorkQueue, DestroyDispatchedOnDispatchQueue)
+        }
+        queue[j] = nullptr;
+    }
+
```
would it be worth checking the refcount of each queues ? in both the loops and at the end that they are all at 1 now (and that all tasks have run)

kkinnunen-apple replied:
The idea is that the refcounts are not 1 and that not all the tasks have run. That's the point of the test:
- to test the scenario where the ref is dropped
- the only ones holding the ref are the multiple pending tasks from within the framework
- the tasks also use the object without the owning ref, as they expect the framework to hold the ref

cdumez reviewed (changes requested):
What's the point of keeping the WorkQueue object alive?

I think we should go the other way and destroy the WorkQueue object as soon as possible unless there is a good reason not to. Our underlying dispatch_queue already ensures that underlying tasks keep running without us having to keep the WorkQueue object alive.

kkinnunen-apple committed:
Cocoa WorkQueue::dispatchWithQOS() and dispatch() work item lifetime semantics do not match
https://bugs.webkit.org/show_bug.cgi?id=260852
rdar://114618174

Reviewed by Jean-Yves Avenard and Chris Dumez.

dispatch() users expect the work function to be destructed in the
queue

Add the same semantic to dispatchWithQOS() to avoid confusion in the
code and to have uniform interface.

* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::dispatchWithQOS):
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/267776@main
